### PR TITLE
Adds std compliant forward iterators for sidre's groups, views, attributes and buffers

### PR DESCRIPTION
**Pull request by kennyweiss**
# Summary

- This PR is a feature that improves sidre's usability
- It adds std-compliant forward iterators for the views and groups within a `Group` and for the attributes and buffers within a `DataStore`
- It does this by reworking the `sidre::ItemCollection` hierarchy, adding iterators to it and exposing them within sidre's `Group` and `DataStore` classes

- The new interface has the following form:

```cpp
for(auto& buff : ds->buffers() )
{
  // do something with the buffers
}

auto* some_grp = ds->getRoot();

for(auto& view : some_grp->views() )
{
  // do something with the views
}

for(auto& grp : some_grp->groups() )
{
  // do something with the groups
}

// alternate form
for(auto it = some_grp->groups().begin(); it != some_grp->groups().end(); ++it)
{
    auto& grp = *it;
    std::string name = it->getName();
    ...
}
```

#### TODO

- [x] Update RELEASE_NOTES
- [x] Document usage in sidre docs
- [x] Highlight API changes (there are very few)
- [x] Add unit tests for `sidre::ItemCollection` and its derived types
- [x] Add iterator adapter for collections to limit access to underlying data

### CODE REVIEW

File: src/axom/core/docs/sphinx/core_acceleration.rst
Comment (white238) [inline src/axom/core/docs/sphinx/core_acceleration.rst:67]: 👍

File: src/axom/sidre/core/DataStore.hpp
Comment (white238) [inline src/axom/sidre/core/DataStore.hpp:191]:
```suggestion
  //!  @name Accessors for buffer collections.
```

File: src/axom/sidre/core/DataStore.hpp
Comment (rhornung67) [inline src/axom/sidre/core/DataStore.hpp:193]:
```suggestion
  //! These methods can be used to iterate over the collection of buffers
```

File: src/axom/sidre/core/DataStore.hpp
Comment (rhornung67) [inline src/axom/sidre/core/DataStore.hpp:203]: Do we really want to expose a non-const reference to the buffer collection?

File: src/axom/sidre/core/DataStore.hpp
Comment (rhornung67) [inline src/axom/sidre/core/DataStore.hpp:390]:
```suggestion
  //!  @name Accessors for attribute collections.
```

File: src/axom/sidre/core/DataStore.hpp
Comment (rhornung67) [inline src/axom/sidre/core/DataStore.hpp:392]:
```suggestion
  //! These methods can be used to iterate over the collection of attributes
```

File: src/axom/sidre/core/Group.cpp
Comment (rhornung67) [inline src/axom/sidre/core/Group.cpp:50]: I find this message and the ones below somewhat unclear. In particular, the "of a (nameless) list Group" part. Thoughts on making it clearer?

File: src/axom/sidre/core/Group.hpp
Comment (rhornung67) [inline src/axom/sidre/core/Group.hpp:898]:
```suggestion
  //!  @name Accessors for group and view collections.
```

File: src/axom/sidre/core/Group.hpp
Comment (rhornung67) [inline src/axom/sidre/core/Group.hpp:900]:
```suggestion
  //! These methods can be used to iterate over the collection of groups and views
```

File: src/axom/sidre/core/Group.hpp
Comment (rhornung67) [inline src/axom/sidre/core/Group.hpp:914]: Expose non-const reference?

File: src/axom/sidre/core/Group.hpp
Comment (rhornung67) [inline src/axom/sidre/core/Group.hpp:924]: Expose non-const ref?

File: src/axom/sidre/core/Group.hpp
Comment (rhornung67) [inline src/axom/sidre/core/Group.hpp:940]: Expose pointer to non-const?

File: src/axom/sidre/core/Group.hpp
Comment (rhornung67) [inline src/axom/sidre/core/Group.hpp:958]: Non-const ptr?

File: src/axom/sidre/core/Group.hpp
Comment (kennyweiss) [reply src/axom/sidre/core/Group.hpp:914]: Good question.  For the range-for, we need something that has `begin()` and `end()` functions. Would you prefer an adapter/view class that hides the rest of the functionality?

File: src/axom/sidre/core/Group.hpp
Comment (kennyweiss) [reply src/axom/sidre/core/Group.hpp:940]: These functions are private, so I think they should be ok.

They're only there to cast from the base type `ItemCollection` to the derived type `MapCollection` (with an assert that it is ok to do so).  The cast is now necessary because I removed functionality related to named items from the base class.

Comment (rhornung67) [reply src/axom/sidre/core/Group.hpp:940]: Sorry. I missed the private part.

File: src/axom/sidre/core/Group.hpp
Comment (rhornung67) [reply src/axom/sidre/core/Group.hpp:914]: It's your call. I want to avoid folks accidentally messing this up.

File: src/axom/sidre/core/ItemCollection.hpp
Comment (kanye-quest) [inline src/axom/sidre/core/ItemCollection.hpp:205]: Is there a reason these have to be forward iterators, and not random-access iterators?

File: src/axom/sidre/tests/sidre_group.cpp
Comment (nselliott) [inline src/axom/sidre/tests/sidre_group.cpp:808]: I think we need a test like this that uses the new iterator for the ListCollection.  I don't see any test coverage for that.

Comment (kennyweiss) [reply src/axom/sidre/tests/sidre_group.cpp:808]: Agreed -- actually, it doesn't look like we have any unit tests for the `ItemCollection` or its derived classes. I'll add some.

**Comment by agcapps** [comment]
If you put refactorings in their own PRs (such as changing `#include "axom/sidre/core/sidre.hpp" // for sidre::Group, sidre::View` to `#include "axom/sidre.hpp"`), it would make the PRs cleaner, quicker, and easier to review.

(says one who has mixed all sorts of things into PRs before)

File: src/axom/sidre/core/MFEMSidreDataCollection.hpp
Comment (agcapps) [inline src/axom/sidre/core/MFEMSidreDataCollection.hpp:12]: Including these three headers in favor of sidre.hpp is just a refactoring, right?  I like this; why did you make this change?

File: src/axom/sidre/core/MapCollection.hpp
Comment (agcapps) [inline src/axom/sidre/core/MapCollection.hpp:62]: Why this change from `TYPE` to `T` throughout?

File: src/axom/sidre/core/MFEMSidreDataCollection.cpp
Comment (kennyweiss) [reply src/axom/sidre/core/MFEMSidreDataCollection.cpp:20]: The code currently (i.e. before this branch) had a hand-written file `sidre/core/sidre.hpp`, which was not up to date. 
I removed that file in favor of the automatically generated `sidre.hpp` for files outside of sidre.
For internal files, I added the few explicit headers, as needed.

File: src/axom/sidre/core/MFEMSidreDataCollection.hpp
Comment (kennyweiss) [reply src/axom/sidre/core/MFEMSidreDataCollection.hpp:12]: (please see my response to a similar comment on `MFEMSidreDataCollection`)

File: src/axom/sidre/core/ItemCollection.hpp
Comment (kanye-quest) [reply src/axom/sidre/core/ItemCollection.hpp:205]: Just asking out of curiosity, although I think there are some STL algorithms that require random-access iterators.

But doesn't sound like it'd be worth it if we can't implement it in a straightforward manner (also, we'd need to do it in O(1) to comply with the standard).

File: src/axom/sidre/core/MapCollection.hpp
Comment (kennyweiss) [reply src/axom/sidre/core/MapCollection.hpp:62]: I modified it before extending the `ItemCollection` hierarchy. It seemed cleaner/more consistent with our `C++` coding conventions.

File: src/axom/sidre/core/Group.hpp
Comment (kennyweiss) [reply src/axom/sidre/core/Group.hpp:914]: Do others want to weigh in on this issue? 

Briefly, the question is about giving non-const access to the underlying `ItemCollection` for the groups and views or to add an adapter that limits what is exposed. 

The range-for syntax only requires having the `begin()` and `end()` functions.

Comment (kennyweiss) [reply src/axom/sidre/core/Group.hpp:914]: I was/am hoping to use the `IndexedCollection` in tribol, so I'd prefer not to cater it too much to our usage here for `sidre::Group`

Comment (kennyweiss) [reply src/axom/sidre/core/Group.hpp:914]: The adapter idea is probably not too much code and should meet our needs

File: src/axom/sidre/tests/sidre_collections.cpp
Comment (kennyweiss) [inline src/axom/sidre/tests/sidre_collections.cpp:8]: Note: sidre's `ItemCollection` hierarchy was previously untested.
This file covers tests for all derived classes of `ItemCollection`

**Comment by agcapps** [comment]
@kennyweiss , I made a request for several more tests.  I think I am close to approving after that.

File: src/axom/sidre/tests/sidre_collections.cpp
Comment (kennyweiss) [reply src/axom/sidre/tests/sidre_collections.cpp:543]: Thanks for the suggestions @agcapps.

I added all the requested tests and improved the consistency between `MapCollection` and `IndexedCollection's handling of collisions and illegal keys (via changes to the new class `IndexedCollection`)

File: src/axom/sidre/core/Group.cpp
Comment (kennyweiss) [reply src/axom/sidre/core/Group.cpp]: Reworded

File: src/axom/sidre/core/Group.hpp
Comment (agcapps) [reply src/axom/sidre/core/Group.hpp:914]: @kennyweiss , do you plan to make an adapter or view class?

Comment (kennyweiss) [reply src/axom/sidre/core/Group.hpp:914]: Working on it... 
Ran into an issue with const/non-const access that I still have to figure out, but getting close

File: src/axom/sidre/tests/sidre_group.cpp
Comment (nselliott) [reply src/axom/sidre/tests/sidre_group.cpp:814]: Thanks, @kennyweiss , the `sidre_collections` test looks good.

File: src/axom/sidre/core/DataStore.hpp
Comment (kennyweiss) [reply src/axom/sidre/core/DataStore.hpp]: I added an `iterator_adaptor` utility class to `ItemCollection` to only expose the functions required for iteration, e.g. `begin()` and `end()`

File: src/axom/sidre/core/Group.hpp
Comment (kennyweiss) [reply src/axom/sidre/core/Group.hpp]: I added an `iterator_adaptor` utility class to `ItemCollection` to only expose the functions required for iteration, e.g. `begin()` and `end()`

**Comment by kennyweiss** [comment]
I've added all requested changes/features to this branch, including improved testing (and bugfixes!) on the `ItemCollection` and an iterator adaptor to limit access/modifications to the collection

File: src/axom/sidre/core/ItemCollection.hpp
Comment (kennyweiss) [inline src/axom/sidre/core/ItemCollection.hpp:143]: Note: I used `adaptor` vs. `adapter` for consistency w/ the C++ std library.

See: https://stackoverflow.com/questions/2076321/adapter-or-adaptor

File: src/axom/sidre/core/ListCollection.hpp
Comment (kennyweiss) [inline src/axom/sidre/core/ListCollection.hpp:16]: @nselliott -- When you get a chance, could you please check if `ListCollection` is still necessary, or if we should favor the new `IndexedCollection`.

The main difference is that `ListCollection` maintains a list of the used indexes, which provides fast O(1) insertions, but O(n) removals.  This is stress testing in `sidre_collections` tests that adds and removes 1000s of items and `ListCollection` is noticeably slower there.

In contrast, `IndexedCollection` has more complex logic to support adding items at a user-supplied index, which was necessary for creating `Buffer`s in the DataStore class.

Comment (nselliott) [reply src/axom/sidre/core/ListCollection.hpp:16]: It looks like `IndexedCollection` satisfies the use case motivations for `ListCollection`, allowing a `Group` to hold a collection of unnamed objects, and iteration that preserves the order of insertion, if the items are inserted without user-specified indices.

What is the performance cost of insertion in `IndexedCollection` compared to `ListCollection`? O(1) is nice to have, but if we're not adding too much to that, it seems a good idea to eliminate one of the classes to avoid redundant functionality.  I think leaving it as a subsequent task and not adding it to this PR would be the way to go, though.

Comment (kennyweiss) [reply src/axom/sidre/core/ListCollection.hpp:16]: Thanks @nselliott -- I agree about swapping it out in a follow-up PR.

Insertion in an `IndexedCollection` is amortized O(1) -- if the index is in range, it will pop the `free_id` stack, otherwise, it grows the size of the array (and fills with `nullptr` to match the index.  The only slightly complicated part is that indices in the `free_ids` list might be occupied, so we need to check `hasItem`  before using them.

**Review by agcapps — APPROVED** [review]
Quite the endeavor, @kennyweiss .  Thank you.

**Review by rhornung67 — APPROVED** [review]
Thanks @kennyweiss good improvements!

**Pull request merged**
